Support fetching secrets

imperfect-fourth opened the pull request:
Closes #1 
Closes #3 

Normally, plugin properties are available to us in the form of env vars of pattern `$BUILDKITE_PLUGIN_<PLUGIN_NAME>_<PROPERTY>`. But this plugin takes input secrets as an array of mappings and buildkite transpiles map objects into this form `$BUILDKITE_PLUGIN_<PLUGIN_NAME>_<PARAMETER_NAME>_<index 0,1,...>_<SUB_PROPERTY>`.

To loop over these env vars, I've had to use something called `variable indirection` in bash([reference](https://www.gnu.org/software/bash/manual/html_node/Shell-Parameter-Expansion.html)). Let me know if there's a cleaner way to do this.

Successful buildkite [run](https://buildkite.com/hasura/smooth-secrets-test/builds/156#79be6139-5494-475f-a436-2d4e10da0f44)

imperfect-fourth committed:
echo input fields

imperfect-fourth committed:
print env in pre-checkout

imperfect-fourth committed:
setup ssh keys in pre-checkout

imperfect-fourth committed:
set pipefail

imperfect-fourth committed:
fix env var names

imperfect-fourth committed:
fix file creation logic

imperfect-fourth committed:
add pre-exit hook

imperfect-fourth committed:
fix export statement

imperfect-fourth committed:
fix possible error

imperfect-fourth committed:
set trap to report error

imperfect-fourth committed:
fix loop increment statement

imperfect-fourth committed:
Fix arithmetic operation

As it turns out, "((expr))" evaluation returns 1 if expr returns 0, so
the previous implementation kept failing because of non zero exit code

scriptnull commented:
Will be using this comment for sharing the outputs of the draft we have so far

```yaml
steps:
  - label: "Buildkite agent pipeline upload"
    plugins:
      - hasura/smooth-secrets#a441f9e6eea4ed4314b7c83d2cd67a4ee30e9815:
    command: ls
```

![image](https://user-images.githubusercontent.com/4211715/129147689-681d9b9d-43db-470d-b099-a8046b441e3d.png)

imperfect-fourth commented:
It seems buildkite itself doesn't enforce the `required` directive in `plugin.yml`. I will add a check for that in the hook.

imperfect-fourth committed:
add readme and restructure hook code

imperfect-fourth committed:
remove keys instead of deleting ssh-agent

imperfect-fourth committed:
add debug statements

imperfect-fourth committed:
add log statement in pre-exit hook

imperfect-fourth committed:
fix log statement

imperfect-fourth committed:
remove exit 0 from hook

the exported variables were not available in the subsequent hooks and
commands since exit 0 was exiting the current shell and a new shell was
being created with a new environment(took way too long to figure this
out)

imperfect-fourth committed:
cleanup

imperfect-fourth committed:
small fix

scriptnull commented on hooks/pre-checkout:
```diff
@@ -0,0 +1,85 @@
+while true; do
+  strategy="$(get_indirected_env "$i" "STRATEGY")"
+
+  if [ -z "$strategy" ]; then
+    break
+  fi
```
Probably remove this check and let it hit the default case in the switch case below?

imperfect-fourth replied:
Since we don't know how many secrets there are in the "array" we can't add this to the default case. The default case will be used to skip any non-supported `strategy` values and the empty case will be used to break out of the loop.

scriptnull replied:
lol :joy: what was I thinking, haha

scriptnull commented on hooks/pre-checkout:
```diff
@@ -0,0 +1,85 @@
+      aws_strategy "$i"
+      ;;
+    *)
+      echo "Strategy \"${strategy}\" not supported"
+      exit 1
```
```bash
echo "skipping to setup secret[i] due to invalid strategy field (${strategy})"
continue
```

If we are trying to setup multiple secrets and if the first one is invalid, then using a `continue` instead of `exit` would help in trying to setup the next secret in the array if it is valid, right?

scriptnull commented on hooks/pre-exit:
```diff
@@ -0,0 +1,10 @@
+
+set -eo pipefail
+
+echo "--- Cleaning up secrets"
```
Needs emoji magic here :smiley:

scriptnull commented on hooks/pre-exit:
```diff
@@ -0,0 +1,10 @@
+echo "--- Cleaning up secrets"
+echo "Killing ssh-agent with PID ${SSH_AGENT_PID}"
+eval "$(ssh-agent -k)"
+echo "Removing secrets directory: ${SECRETS_DIR}"
```
Shouldn't we expose this env via an environment hook in order for it to be accessible here?

scriptnull commented on plugin.yml:
```diff
@@ -0,0 +1,23 @@
+configuration:
+  properties:
+    secrets:
+      type: object
```
we should probably denote this to be an array of objects with the following properties

scriptnull commented on hooks/pre-checkout:
```diff
@@ -0,0 +1,85 @@
+
+  case "${secret_type}" in
+    "ssh")
+      echo "Fetching secret from secrets manager"
```
`secrets manager` -> `AWS secrets manager`

scriptnull committed:
enable linting

imperfect-fourth committed:
cleanup

imperfect-fourth committed:
minor fix

imperfect-fourth committed:
fix lint errors

scriptnull commented on plugin.yml:
```diff
@@ -0,0 +1,25 @@
+  properties:
+    secrets:
+      type: array
+      properties:
+        strategy:
+          type: string
+        key:
+          type: string
+        type:
+          type: string
+        region:
+          type: string
+      required:
+        - strategy
+        - key
+        - type
+        - region
```
I think we should be using the `items` property here and move this object definition inside that. https://json-schema.org/understanding-json-schema/reference/array.html#id8

(at the same time I wonder why json scehma validation isn't being done on buildkite's side :thinking: )

scriptnull commented on hooks/pre-exit:
```diff
@@ -0,0 +1,10 @@
+echo "Killing ssh-agent with PID ${SSH_AGENT_PID}"
+eval "$(ssh-agent -k)"
+echo "Removing secrets directory: ${SECRETS_DIR}"
+[ -d "${SECRETS_DIR}" ] && rm -r "${SECRETS_DIR}"
```
Why not `rm -rf "$SECRETS_DIR"` ?

imperfect-fourth committed:
fix json schema

imperfect-fourth committed:
support env secret type

imperfect-fourth committed:
fix errors

imperfect-fourth committed:
add encoding field

imperfect-fourth committed:
add env example in readme

scriptnull committed:
docs refactor

imperfect-fourth committed:
fix conditional

scriptnull committed:
Create README.md

imperfect-fourth committed:
add check for required fields
